go-controller/pkg/clustermanager/secondary_network_cluster_controller.go

```diff
@@ -0,0 +1,321 @@
+}
+
+// Stop gracefully stops the controller, and delete all logical entities for this network if requested
+func (sncc *secondaryNetworkClusterController) Stop() {
+	klog.Infof("Stop secondary %s network cluster manager for network %s", sncc.TopologyType(), sncc.GetNetworkName())
+	close(sncc.stopChan)
+	sncc.wg.Wait()
+
+	if sncc.nodeHandler != nil {
+		sncc.watchFactory.RemoveNodeHandler(sncc.nodeHandler)
+	}
+}
+
+// Cleanup cleans up logical entities for the given network, called from net-attach-def routine
+func (sncc *secondaryNetworkClusterController) Cleanup(netName string) error {
+	return sncc.nodeL3NetworkController.cleanup(netName)
+}
+
+// Run starts the actual watching.
+func (sncc *secondaryNetworkClusterController) Run() error {
+	if err := sncc.WatchNodes(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// WatchNodes starts the watching of node resource and calls
+// back the appropriate handler logic
+func (sncc *secondaryNetworkClusterController) WatchNodes() error {
+	if sncc.nodeHandler != nil {
+		return nil
+	}
+	handler, err := sncc.retryNodes.WatchResource()
+	if err == nil {
+		sncc.nodeHandler = handler
+	}
+	return err
+}
+
+func (sncc *secondaryNetworkClusterController) addUpdateNodeEvent(node *kapi.Node, nSyncs *nodeSyncs) error {
+	klog.Infof("Adding or Updating Node %q for network %s", node.Name, sncc.GetNetworkName())
+	if nSyncs.syncNode {
+		if err := sncc.addNode(node); err != nil {
+			sncc.addNodeFailed.Store(node.Name, true)
+			err = fmt.Errorf("nodeAdd: error adding node %q for network %s: %w", node.Name, sncc.GetNetworkName(), err)
+			return err
+		}
+		sncc.addNodeFailed.Delete(node.Name)
+	}
+
+	return nil
+}
+
+func (sncc *secondaryNetworkClusterController) addNode(node *kapi.Node) error {
+	return sncc.nodeL3NetworkController.addUpdateNode(node)
+}
+
+func (sncc *secondaryNetworkClusterController) deleteNodeEvent(node *kapi.Node) error {
+	klog.V(5).Infof("Deleting Node %q for network %s. Removing the node from "+
+		"various caches", node.Name, sncc.GetNetworkName())
+
+	return sncc.nodeL3NetworkController.deleteNode(node)
+}
+
+func (sncc *secondaryNetworkClusterController) syncNodes(nodes []interface{}) error {
+	return sncc.nodeL3NetworkController.syncNodes(nodes)
+}
+
+// newRetryFramework builds and returns a retry framework for the input resource type;
+func (sncc *secondaryNetworkClusterController) newRetryFramework(
+	objectType reflect.Type) *retry.RetryFramework {
+	eventHandler := &secondaryNetworkClusterControllerEventHandler{
+		objType:      objectType,
+		watchFactory: sncc.watchFactory,
+		sncc:         sncc,
+		syncFunc:     nil,
+	}
+	resourceHandler := &retry.ResourceHandler{
+		HasUpdateFunc:          true,
+		NeedsUpdateDuringRetry: false,
+		ObjType:                objectType,
+		EventHandler:           eventHandler,
+	}
+	return retry.NewRetryFramework(
+		sncc.stopChan,
+		sncc.wg,
+		sncc.watchFactory,
+		resourceHandler,
+	)
+}
+
+type nodeSyncs struct {
```

Comment:
Why do this? why not just pass a bool into funcs? Seems redundant to wrap a struct around the bool.